risc0/zkp/src/hal/metal.rs

```diff
@@ -45,16 +50,136 @@ const KERNEL_NAMES: &[&str] = &[
+        &self,
+        hal: &MetalHal<Self>,
+        output: &BufferImpl<Digest>,
+        matrix: &BufferImpl<BabyBearElem>,
+    ) {
+        let row_size = output.size();
+        let col_size = matrix.size() / output.size();
+        assert_eq!(matrix.size(), col_size * row_size);
+        let args = &[
+            output.as_arg(),
+            matrix.as_arg(),
+            KernelArg::Integer(row_size as u32),
+            KernelArg::Integer(col_size as u32),
+        ];
+        hal.dispatch_by_name("sha_rows", args, row_size as u64);
+    }
+}
+
+pub struct MetalHashPoseidon {
+    round_constants: BufferImpl<BabyBearElem>,
+    mds: BufferImpl<BabyBearElem>,
+    partial_comp_matrix: BufferImpl<BabyBearElem>,
+    partial_comp_offset: BufferImpl<BabyBearElem>,
+}
+
+impl MetalHash for MetalHashPoseidon {
+    type HashSuite = HashSuitePoseidon;
+
+    fn new(hal: &MetalHal<Self>) -> Self {
+        use crate::core::poseidon;
+        let round_constants = hal.copy_from_elem("round_constants", poseidon::ROUND_CONSTANTS);
+        let mds = hal.copy_from_elem("mds", poseidon::MDS);
+        let partial_comp_matrix =
+            hal.copy_from_elem("partial_comp_matrix", &*poseidon::PARTIAL_COMP_MATRIX);
+        let partial_comp_offset =
+            hal.copy_from_elem("partial_comp_offset", &*poseidon::PARTIAL_COMP_OFFSET);
+        MetalHashPoseidon {
+            round_constants,
+            mds,
+            partial_comp_matrix,
+            partial_comp_offset,
+        }
+    }
+
+    fn hash_fold(&self, hal: &MetalHal<Self>, io: &BufferImpl<Digest>, output_size: usize) {
+        let args = &[
+            self.round_constants.as_arg(),
+            self.mds.as_arg(),
+            self.partial_comp_matrix.as_arg(),
+            self.partial_comp_offset.as_arg(),
+            io.as_arg_with_offset(output_size),
+            io.as_arg_with_offset(output_size * 2),
+        ];
+        hal.dispatch_by_name("poseidon_fold", args, output_size as u64);
+    }
+
+    fn hash_rows(
+        &self,
+        hal: &MetalHal<Self>,
+        output: &BufferImpl<Digest>,
+        matrix: &BufferImpl<BabyBearElem>,
+    ) {
+        let row_size = output.size();
+        let col_size = matrix.size() / output.size();
+        assert_eq!(matrix.size(), col_size * row_size);
+        let args = &[
+            self.round_constants.as_arg(),
+            self.mds.as_arg(),
+            self.partial_comp_matrix.as_arg(),
+            self.partial_comp_offset.as_arg(),
+            output.as_arg(),
+            matrix.as_arg(),
+            KernelArg::Integer(row_size as u32),
+            KernelArg::Integer(col_size as u32),
+        ];
+        hal.dispatch_by_name("poseidon_rows", args, row_size as u64);
+    }
+}
+
 #[derive(Debug)]
-pub struct MetalHal {
+pub struct MetalHal<Hash: MetalHash + ?Sized> {
     pub device: Device,
     pub cmd_queue: CommandQueue,
     kernels: HashMap<String, ComputePipelineDescriptor>,
+    pub library: Library,
```

Comment:
Is this still needed? I'm not immediately seeing who uses this externally?

Reply:
So I already know third parties are interested in building their own hashes, so I was considering that use case.  Although, I suppose they don't probably need access to the metal library itself, so I'll drop.